notebooks/implicit_differentiation.ipynb

```diff
@@ -0,0 +1,561 @@
+    "    while cond_fn(state, steps):\n",
+    "        nbrs = update_nbrs(state.position,nbrs)\n",
+    "        new_state = apply_fn(state, neighbor=nbrs)\n",
+    "        if forced_rebuilding and steps == 10:\n",
+    "            print(\"Forced rebuilding of neighbor_list.\")\n",
+    "            nbrs = neighbor_fn.allocate(state.position)\n",
+    "        if nbrs.did_buffer_overflow:\n",
+    "            print(\"Rebuilding neighbor_list.\")\n",
+    "            nbrs = neighbor_fn.allocate(state.position)\n",
+    "        else:\n",
+    "            state = new_state\n",
+    "            steps += 1\n",
+    "        \n",
+    "    return state.position, nbrs, steps+1\n",
+    "\n",
+    "# This version is fully differentiable and jit-able\n",
+    "def run_minimization_scan(force_fn, R_init, shift, num_steps=5000, **kwargs):\n",
+    "    init, apply = minimize.fire_descent(jit(force_fn), shift, **kwargs)\n",
+    "    apply = jit(apply)\n",
+    "\n",
+    "    @jit\n",
+    "    def scan_fn(state, i):\n",
+    "        return apply(state), 0.0\n",
+    "\n",
+    "    state = init(R_init)\n",
+    "    state, _ = lax.scan(scan_fn, state, jnp.arange(num_steps))\n",
+    "    return state.position, jnp.amax(jnp.abs(force_fn(state.position)))"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "introductory-priority",
+   "metadata": {},
+   "source": [
+    "# Meta Optimization"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "industrial-tonight",
+   "metadata": {},
+   "source": [
+    "In this notebook we'll have another look at differentiating through an energy minimization routine. But this time we'll look at how a technique called implicit differentiation lets us do so much more efficiently than before. Let us first set up our system and see for ourselves what goes wrong when we aren't using implicit differentiation."
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "liable-dodge",
+   "metadata": {},
+   "source": [
+    "We'll work with a system of `N` soft spheres and we are interested in first computing the energy minimum of this system and then to compute the gradient of the energy with respect to the parameters `sigma` and/or `alpha`."
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "wanted-consequence",
+   "metadata": {},
+   "source": [
+    "We'll start of with a very small and rather dilute system."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "creative-assault",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "N = 7\n",
+    "density = 0.3\n",
+    "\n",
+    "dimension = 2\n",
+    "box_size = quantity.box_size_at_number_density(N,density,dimension)\n",
+    "displacement, shift = space.periodic(box_size) \n",
+    "\n",
+    "key = random.PRNGKey(5)\n",
+    "key, split = random.split(key)\n",
+    "R_init = random.uniform(key, (N,dimension), minval=0.0, maxval=box_size, dtype=f64) \n",
+    "\n",
+    "sigma = jnp.full((N,N), 2.)\n",
```

Comment:
Can you add spaces between commas and the next character `(N, N)` rather than `(N,N)`.